pkg/check/zones.go

```diff
@@ -0,0 +1,189 @@
+		referencesToCheck = append(referencesToCheck, ancestor.Reference())
+	}
+	attachedTags, err := getAttachedTagsOnObjects(ctx, referencesToCheck)
+	if err != nil {
+		klog.Error("Unable to get attached tags.")
+		return err
+	}
+
+	klog.V(4).Infof("Processing attached tags")
+	regionTagAttached := false
+	zoneTagAttached := false
+	for _, attachedTag := range attachedTags {
+		for _, tag := range attachedTag.Tags {
+			klog.V(5).Infof("Current tag: %s", tag)
+			if !regionTagAttached {
+				if tag.CategoryID == vctx.regionTagCategoryID {
+					regionTagAttached = true
+					klog.V(4).Infof("Found Region: %s", tag.Name)
+				}
+			}
+			if !zoneTagAttached {
+				if tag.CategoryID == vctx.zoneTagCategoryID {
+					zoneTagAttached = true
+					klog.V(4).Infof("Found Zone: %s", tag.Name)
+				}
+			}
+			if regionTagAttached && zoneTagAttached {
+				return nil
+			}
+		}
+	}
+
+	klog.V(4).Infof("Region %t Zone %t", regionTagAttached, zoneTagAttached)
+	var errs []string
+	if !regionTagAttached {
+		klog.Warning("Region not found")
+		errs = append(errs, fmt.Sprintf("tag associated with tag category %s not attached to this resource or ancestor", TagCategoryRegion))
+	}
+	if !zoneTagAttached {
+		klog.Warning("Zone not found")
+		errs = append(errs, fmt.Sprintf("tag associated with tag category %s not attached to this resource or ancestor", TagCategoryZone))
+	}
+	return errors.New(strings.Join(errs, ","))
+}
+
+// CheckZoneTags will attempt to validate that the necessary tags are present to represent the
+// various zones defined for a cluster.
+func CheckZoneTags(ctx *CheckContext) error {
+	klog.Info("Checking tags for multi-zone support.")
+	var errs []error
+
+	// Get all failure domains
+	klog.V(4).Info("Getting infrastructure configuration.")
+	inf, err := ctx.KubeClient.GetInfrastructure(ctx.Context)
+	if err != nil {
+		klog.Errorf("Error getting infrastructure: %v", err)
+		return err
+	}
+
+	// Perform check if FailureDomains defined.  We need 2 or more to require tags.
+	klog.V(4).Info("Checking failure domains.")
+	if fds := inf.Spec.PlatformSpec.VSphere.FailureDomains; fds != nil && len(fds) > 1 {
+		// Validate tags exist for cluster
+		regionTagCategoryId, zoneTagCategoryId, err := validateTagCategories(ctx)
+		if err != nil {
+			return fmt.Errorf("Multi-Zone support: %s", err)
+		}
+		klog.V(4).Infof("Region: %s  Zone: %s", regionTagCategoryId, zoneTagCategoryId)
+
+		// Iterate through each FailureDomain and check tags.
+		for _, fd := range fds {
+
+			// Validate compute cluster is defined correctly
+			vsphereField := field.NewPath("platform").Child("vsphere")
+			topologyField := vsphereField.Child("failureDomains").Child("topology")
+
+			computeCluster := fd.Topology.ComputeCluster
+			clusterPathRegexp := regexp.MustCompile("^\\/(.*?)\\/host\\/(.*?)$")
+			clusterPathParts := clusterPathRegexp.FindStringSubmatch(computeCluster)
+			if len(clusterPathParts) < 3 {
+				klog.V(4).Info("CLUSTER PARTS LESS THAN 3")
```

Comment:
No all caps logging please.

Reply:
Thanks, I'll fix that.  I had gotten it from somewhere else that way and assumed it was ok.